contracts/external/cover/CoverOrder.sol

```diff
@@ -0,0 +1,310 @@
+import "../../interfaces/INXMMaster.sol";
+import "../../interfaces/INXMToken.sol";
+import "../../interfaces/IWeth.sol";
+import "../../interfaces/IPool.sol";
+
+contract CoverOrder is ICoverOrder, Ownable, EIP712 {
+  using ECDSA for bytes32;
+  using SafeERC20 for IERC20;
+
+  /* ========== STATE VARIABLES ========== */
+  mapping(bytes32 => OrderStatus) public orderStatus;
+
+  /* ========== IMMUTABLES ========== */
+  ICover public immutable cover;
+  IMemberRoles public immutable memberRoles;
+  INXMToken public immutable nxmToken;
+  INXMMaster public immutable master;
+  IWeth public immutable weth;
+
+  /* ========== CONSTANTS ========== */
+  address public constant ETH = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+  uint private constant ETH_ASSET_ID = 0;
+  uint private constant NXM_ASSET_ID = type(uint8).max;
+
+  bytes32 private constant EXECUTE_ORDER_TYPEHASH = keccak256(
+    abi.encodePacked(
+      "ExecuteOrder(",
+      "uint24 productId,",
+      "uint96 amount,",
+      "uint32 period,",
+      "uint8 paymentAsset,",
+      "uint8 coverAsset,",
+      "address owner,",
+      "ExecutionDetails executionDetails)",
+      "ExecutionDetails(uint256 notBefore,uint256 deadline,uint256 maxPremiumInAsset)"
+    )
+  );
+
+  /* ========== CONSTRUCTOR ========== */
+  constructor(
+    address _master,
+    address _weth,
+    address _owner
+  ) EIP712("NexusMutualCoverOrder", "1") {
+    master = INXMMaster(_master);
+    cover = ICover(master.getLatestAddress("CO"));
+    memberRoles = IMemberRoles(master.getLatestAddress("MR"));
+    nxmToken = INXMToken(master.tokenAddress());
+    weth = IWeth(_weth);
+    transferOwnership(_owner);
+  }
+
+  /// @notice Verifies and executes the order to buy cover on behalf of the creator of limit order
+  /// @notice Function only allows users to pay with coverAsset or NXM, this is being checked in the Cover contract
+  /// @param params Cover buy parameters
+  /// @param poolAllocationRequests Pool allocations for the cover
+  /// @param executionDetails Start and end date when the order can be executed and max premium in asset
+  /// @param signature The signature of the order
+  function executeOrder(
+    BuyCoverParams calldata params,
+    PoolAllocationRequest[] calldata poolAllocationRequests,
+    ExecutionDetails calldata executionDetails,
+    bytes calldata signature
+  ) external payable returns (uint coverId) {
+
+    if (block.timestamp > executionDetails.deadline) {
+      revert OrderExpired();
+    }
+
+    if (block.timestamp < executionDetails.notBefore) {
+      revert OrderCannotBeExecutedYet();
+    }
+
+    if (params.maxPremiumInAsset > executionDetails.maxPremiumInAsset) {
+      revert OrderPriceNotMet();
+    }
+
+    if(!memberRoles.checkRole(msg.sender, uint(IMemberRoles.Role.Member))) {
+      revert NotAMember();
+    }
+
+    if (params.owner == address(0) || params.owner == address(this)) {
+      revert InvalidOwnerAddress();
+    }
```

Comment:
is it intentional that we are letting any member execute any other member's order? 

or do we want to restrict this to just be the owner?

Reply:
there might be front-run scenario here if we let any member execute any other member's order.

a `cancelOrder` from a member might be front-run by an attacker with `executeOrder`.

Reply:
Then the limit order looses it's point, if the signer has to come and execute it when the time comes for it. The idea is someone creates an order anyone can execute it when the terms are met. You can always cancel before that.